File: src/lib/workflow.ts
import { invoke } from '@tauri-apps/api/core';
import { z } from 'zod';
import type { Task, Project } from './tauri';

export type WorkTask = Task & {
  assigned_member_id: string | null;
  milestone_id: string | null;
  due_at: string | null;
  tags_json: string;
  is_archived: boolean;
  completed_at: string | null;
};
export type Milestone = {
  id: string;
  project_id: string;
  name: string;
  is_closed: boolean;
  created_at: string;
  total: number;
  completed: number;
};
export type Activity = {
  id: number;
  entity_id: string;
  activity_type: string;
  created_at: string;
};
export type TaskFilter = {
  project_id: string;
  search: string;
  status: string | null;
  member_id: string | null;
  archived: boolean;
};
export type TaskPage = {
  tasks: WorkTask[];
  total: number;
  completed: number;
  project_total: number;
};
export type TaskDetails = {
  id: string;
  project_id: string;
  assigned_member_id: string | null;
  milestone_id: string | null;
  due_at: string | null;
  tags: string[];
};
export const taskSchema = z.object({
  title: z.string().trim().min(1).max(200),
  description: z.string().max(2000),
});
export const tagsSchema = z.array(z.string().trim().min(1).max(40)).max(10);
export const statusLabels: Record<Task['status'], string> = {
  todo: '○ Yapılacak',
  in_progress: '▶ Devam ediyor',
  review: '◇ İncelemede',
  blocked: '! Engelli',
  done: '✓ Tamamlandı',
};
export const priorityLabels: Record<Task['priority'], string> = {
  low: 'Düşük',
  normal: 'Normal',
  high: 'Yüksek',
  critical: 'Kritik',
};
export function isStatus(value: string): value is Task['status'] {
  return Object.hasOwn(statusLabels, value);
}
export function isPriority(value: string): value is Task['priority'] {
  return Object.hasOwn(priorityLabels, value);
}
export const taskPage = (filter: TaskFilter, offset: number) =>
  invoke<TaskPage>('task_page', { filter: { ...filter, offset } });
export const moveTask = (projectId: string, id: string, direction: number) =>
  invoke<void>('move_task', { projectId, id, direction });
export const archiveTask = (projectId: string, id: string, archived: boolean) =>
  invoke<void>('archive_task', { projectId, id, archived });
export const deleteTask = (projectId: string, id: string) =>
  invoke<void>('delete_task', { projectId, id });
export const listMilestones = (projectId: string) =>
  invoke<Milestone[]>('list_milestones', { projectId });
export const createMilestone = (projectId: string, name: string) =>
  invoke<void>('create_milestone', { projectId, name });
export const closeMilestone = (
  projectId: string,
  id: string,
  closed: boolean,
) => invoke<void>('close_milestone', { projectId, id, closed });
export const listActivity = (projectId: string, offset: number) =>
  invoke<Activity[]>('list_activity', { projectId, offset });
export const exportProject = (projectId: string) =>
  invoke<string>('export_project', { projectId });
export const importProject = (json: string) =>
  invoke<Project>('import_project', { json });
export type SaveTask = Omit<TaskDetails, 'id'> & {
  id: string | null;
  title: string;
  description: string;
  status: Task['status'];
  priority: Task['priority'];
};
export const saveWorkTask = (request: SaveTask) =>
  invoke<WorkTask>('save_work_task', { request });

File: src-tauri/src/workflow.rs
#[path = "workflow_repository.rs"]
pub mod repository;
#[path = "workflow_service.rs"]
pub mod service;
use crate::{AppError, AppState};
use chrono::{NaiveDate, Utc};
use serde::{Deserialize, Serialize};
pub use service::save;
use sqlx::{FromRow, SqlitePool};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
#[serde(deny_unknown_fields)]
pub struct WorkTask {
    pub id: String,
    pub project_id: String,
    pub title: String,
    pub description: Option<String>,
    pub status: String,
    pub priority: String,
    pub position: i64,
    pub created_at: String,
    pub updated_at: String,
    pub assigned_member_id: Option<String>,
    pub milestone_id: Option<String>,
    pub due_at: Option<String>,
    pub tags_json: String,
    pub is_archived: bool,
    pub completed_at: Option<String>,
}
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
#[serde(deny_unknown_fields)]
pub struct Milestone {
    pub id: String,
    pub project_id: String,
    pub name: String,
    pub is_closed: bool,
    pub created_at: String,
}
#[derive(Serialize, FromRow)]
pub struct MilestoneProgress {
    pub id: String,
    pub project_id: String,
    pub name: String,
    pub is_closed: bool,
    pub created_at: String,
    pub total: i64,
    pub completed: i64,
}
#[derive(Debug, Serialize, Deserialize, FromRow)]
#[serde(deny_unknown_fields)]
pub struct Activity {
    pub id: i64,
    pub project_id: String,
    pub entity_id: String,
    pub activity_type: String,
    pub created_at: String,
}
#[derive(Debug, Deserialize)]
pub struct TaskFilter {
    pub project_id: Uuid,
    pub search: String,
    pub status: Option<String>,
    pub member_id: Option<Uuid>,
    pub archived: bool,
    pub offset: i64,
}
#[derive(Serialize)]
pub struct TaskPage {
    pub tasks: Vec<WorkTask>,
    pub total: i64,
    pub completed: i64,
    pub project_total: i64,
}
#[derive(Debug, Deserialize)]
pub struct TaskDetails {
    pub assigned_member_id: Option<Uuid>,
    pub milestone_id: Option<Uuid>,
    pub due_at: Option<String>,
    pub tags: Vec<String>,
}

#[derive(Deserialize)]
pub struct SaveTask {
    pub id: Option<Uuid>,
    pub project_id: Uuid,
    pub title: String,
    pub description: Option<String>,
    pub status: String,
    pub priority: String,
    pub assigned_member_id: Option<Uuid>,
    pub milestone_id: Option<Uuid>,
    pub due_at: Option<String>,
    pub tags: Vec<String>,
}

#[tauri::command]
pub async fn save_work_task(
    state: tauri::State<'_, AppState>,
    request: SaveTask,
) -> Result<WorkTask, AppError> {
    save(&state.pool, request).await
}

#[tauri::command]
pub async fn task_page(
    state: tauri::State<'_, AppState>,
    filter: TaskFilter,
) -> Result<TaskPage, AppError> {
    repository::page(&state.pool, filter).await
}

#[tauri::command]
pub async fn move_task(
    state: tauri::State<'_, AppState>,
    project_id: Uuid,
    id: Uuid,
    direction: i64,
) -> Result<(), AppError> {
    service::move_task(&state.pool, project_id, id, direction).await
}
#[tauri::command]
pub async fn archive_task(
    state: tauri::State<'_, AppState>,
    project_id: Uuid,
    id: Uuid,
    archived: bool,
) -> Result<(), AppError> {
    service::archive(&state.pool, project_id, id, archived).await
}
#[tauri::command]
pub async fn delete_task(
    state: tauri::State<'_, AppState>,
    project_id: Uuid,
    id: Uuid,
) -> Result<(), AppError> {
    service::delete(&state.pool, project_id, id).await
}
#[tauri::command]
pub async fn list_milestones(
    state: tauri::State<'_, AppState>,
    project_id: Uuid,
) -> Result<Vec<MilestoneProgress>, AppError> {
    repository::milestones(&state.pool, project_id).await
}
#[tauri::command]
pub async fn create_milestone(
    state: tauri::State<'_, AppState>,
    project_id: Uuid,
    name: String,
) -> Result<(), AppError> {
    service::milestone(&state.pool, project_id, name).await
}
#[tauri::command]
pub async fn close_milestone(
    state: tauri::State<'_, AppState>,
    project_id: Uuid,
    id: Uuid,
    closed: bool,
) -> Result<(), AppError> {
    service::close_milestone(&state.pool, project_id, id, closed).await
}
#[tauri::command]
pub async fn list_activity(
    state: tauri::State<'_, AppState>,
    project_id: Uuid,
    offset: i64,
) -> Result<Vec<Activity>, AppError> {
    repository::activity(&state.pool, project_id, offset).await
}
